lib/scholar/cluster/affinity_propagation.ex

```diff
@@ -307,14 +308,12 @@ defmodule Scholar.Cluster.AffinityPropagation do
   end
 
   defnp initialize_similarities(data, opts \\ []) do
-    {n, dims} = Nx.shape(data)
+    n = Nx.axis_size(data, 0)
     self_preference = opts[:self_preference]
-    t1 = Nx.reshape(data, {1, n, dims}) |> Nx.broadcast({n, n, dims})
-    t2 = Nx.reshape(data, {n, 1, dims}) |> Nx.broadcast({n, n, dims})
 
-    dist =
-      (-1 * Scholar.Metrics.Distance.squared_euclidean(t1, t2, axes: [-1]))
-      |> Nx.as_type(to_float_type(data))
+    norm1 = Nx.sum(data ** 2, axes: [1], keep_axes: true)
+    norm2 = Nx.transpose(norm1)
```

Comment:
Is is squared_euclidean so expensive? Is there something we should do about it?

Reply:
In this specific case, it is very expensive because we need to create first a tensor of dim `{num_samples, num_sample, num_features}`. So I wrote a formula that calculates it without that expensive broadcasting. Is it something we should do about it? Hmm, rather no. Our `squared_euclidean` is more general than calculating only pairwise distance. What we can do is actually support pairwise distances in general. They can be calculated using a general formula:
![image](https://github.com/elixir-nx/scholar/assets/56299341/f8757380-17a3-422e-97a1-a4ba7d0e13e7)
(This one is actually for squared distance, but it can be simply wrapped into sqrt). This `|| x ||` is literally any norm and because all norms induces metrics (all normed vector spaces are metric spaces) we can implement any pairwise distance this way :)

Reply:
Excellent, thank you!

Reply:
@josevalim do we plan to add pairwise distances? If yes, then I'll create an issue.

Reply:
Your call!